fastly/base_fastly_service_v1.go

```diff
@@ -0,0 +1,482 @@
+
+			// Active Version represents the currently activated version in Fastly. In
+			// Terraform, we abstract this number away from the users and manage
+			// creating and activating. It's used internally, but also exported for
+			// users to see.
+			"active_version": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			// Cloned Version represents the latest cloned version by the provider. It
+			// gets set whenever Terraform detects changes and clones the currently
+			// activated version in order to modify it. Active Version and Cloned
+			// Version can be different if the Activate field is set to false in order
+			// to prevent the service from being activated. It is not used internally,
+			// but it is exported for users to see after running `terraform apply`.
+			"cloned_version": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+
+			"activate": {
+				Type:        schema.TypeBool,
+				Description: "Conditionally prevents the Service from being activated",
+				Default:     true,
+				Optional:    true,
+			},
+
+			"force_destroy": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+		},
+	}
+
+	for _, a := range serviceDef.GetAttributeHandler() {
+		a.Register(s) // Mutates s, adding handler-specific schema items to the list
+	}
+
+	return s
+}
+
+// resourceCreate satisfies the Terraform resource schema Create "interface"
+// while injecting the ServiceDefinition into the true Create functionality
+func resourceCreate(serviceDef ServiceDefinition) schema.CreateFunc {
+	return func(data *schema.ResourceData, i interface{}) error {
+		return resourceServiceCreate(data, i, serviceDef)
+	}
+}
+
+// resourceRead satisfies the Terraform resource schema Read "interface"
+// while injecting the ServiceDefinition into the true Read functionality
+func resourceRead(serviceDef ServiceDefinition) schema.ReadFunc {
+	return func(data *schema.ResourceData, i interface{}) error {
+		return resourceServiceRead(data, i, serviceDef)
+	}
+}
+
+// resourceUpdate satisfies the Terraform resource schema Update "interface"
+// while injecting the ServiceDefinition into the true Update functionality
+func resourceUpdate(serviceDef ServiceDefinition) schema.UpdateFunc {
+	return func(data *schema.ResourceData, i interface{}) error {
+		return resourceServiceUpdate(data, i, serviceDef)
+	}
+}
+
+// resourceDelete satisfies the Terraform resource schema Delete "interface"
+// while injecting the ServiceDefinition into the true Delete functionality
+func resourceDelete(serviceDef ServiceDefinition) schema.DeleteFunc {
+	return func(data *schema.ResourceData, i interface{}) error {
+		return resourceServiceDelete(data, i, serviceDef)
+	}
+}
+
+// resourceServiceCreate provides service resource Create functionality
+func resourceServiceCreate(d *schema.ResourceData, meta interface{}, serviceDef ServiceDefinition) error {
+	if err := validateVCLs(d); err != nil {
+		return err
+	}
+
+	conn := meta.(*FastlyClient).conn
```

Comment:
Isn't `meta.(*FastlyClient)` suggesting that `meta` is a `*FastlyClient`?

Ah I see. Okay this makes sense.

Note: I know that the `interface{}` is to satisfy the Terraform interface.

**Something to think about and that might not work.**

I wonder if at this level, we could have a function signature like the following:

```go
type Meta struct {
    Conn *FastlyClient
}

func resourceServiceCreate(d *schema.ResourceData, meta Meta, serviceDef ServiceDefinition) error { ... }
```

This way the function signature wouldn't change if we needed to pass additional meta data downstream. We could do the interface assertion above too.

Reply:
Actually, maybe `meta` is just `conn FastlyClient`.

I would rather have the function signature change when we need to add additional dependencies rather than a `Meta` struct that hides a lot of information.

Even if we did go the `Meta` route, I would still like to see `FastlyClient` separate to indicate that it makes external calls.